google/cloud/bigtable/test_proxy/cbt_test_proxy.cc

```diff
@@ -0,0 +1,398 @@
+    new_cell->set_value(std::move(cell.value()));
+  }
+  return new_row;
+}
+
+}  // namespace
+
+grpc::Status CbtTestProxy::CreateClient(
+    ::grpc::ServerContext*, testpb::CreateClientRequest const* request,
+    testpb::CreateClientResponse*) {
+  GCP_LOG(DEBUG) << "searching for: " << request->client_id();
+  if (request->client_id().empty() || request->data_target().empty()) {
+    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
+                        "both `client_id` and `data_target` must be provided.");
+  }
+
+  auto options =
+      Options{}
+          .set<DataEndpointOption>(request->data_target())
+          .set<GrpcCredentialOption>(grpc::InsecureChannelCredentials())
+          .set<MaxConnectionRefreshOption>(std::chrono::milliseconds(0))
+          .set<AppProfileIdOption>(request->app_profile_id());
+
+  if (request->has_per_operation_timeout()) {
+    auto duration = absl::ToChronoMilliseconds(absl::Milliseconds(
+        google::protobuf::util::TimeUtil::DurationToMilliseconds(
+            request->per_operation_timeout())));
+    options
+        .set<DataRetryPolicyOption>(
+            DataLimitedTimeRetryPolicy(duration).clone())
+        // TODO(#4926) - set deadlines using a nicer API.
+        .set<google::cloud::internal::GrpcSetupOption>(
+            [duration](grpc::ClientContext& context) {
+              auto deadline = std::chrono::system_clock::now() + duration;
+              if (context.deadline() >= deadline) {
+                context.set_deadline(deadline);
+              }
+            });
+  }
+
+  std::lock_guard<std::mutex> lk(mu_);
+  auto client = connections_.find(request->client_id());
+  if (client != connections_.end()) {
+    return grpc::Status(
+        grpc::StatusCode::ALREADY_EXISTS,
+        absl::StrCat("Client ", request->client_id(), " already exists."));
+  }
+  connections_.insert({request->client_id(), MakeDataConnection(options)});
+  return grpc::Status();
+}
+
+grpc::Status CbtTestProxy::CloseClient(
+    ::grpc::ServerContext*, testpb::CloseClientRequest const* request,
+    testpb::CloseClientResponse*) {
+  return grpc::Status();
+}
+
+grpc::Status CbtTestProxy::RemoveClient(
+    ::grpc::ServerContext*, testpb::RemoveClientRequest const* request,
+    testpb::RemoveClientResponse*) {
+  std::lock_guard<std::mutex> lk(mu_);
+  auto client = connections_.find(request->client_id());
+  if (client == connections_.end()) {
+    return grpc::Status(
+        grpc::StatusCode::NOT_FOUND,
+        absl::StrCat("Client ", request->client_id(), " not found."));
+  }
+
+  connections_.erase(request->client_id());
+  return grpc::Status();
+}
+
+grpc::Status CbtTestProxy::ReadRow(::grpc::ServerContext*,
+                                   testpb::ReadRowRequest const* request,
+                                   testpb::RowResult* response) {
+  auto sor = GetTableFromRequest(request->client_id(), request->table_name());
```

Comment:
The naming conventions in `google-cloud-cpp` would call this `table` and we would just use it via `table->`.

Reply:
Done.